lib/active_model/serializer/adapter/json.rb

```diff
@@ -4,44 +4,134 @@ class ActiveModel::Serializer::Adapter::Json < ActiveModel::Serializer::Adapter
+            key = association.key
+
+            # sanity check if the association has nesting data
+            has_nesting = nested_associations[key].present?
+            if has_nesting
+              include_options_from_parent = { include: nested_associations[key] }
+              opts = opts.merge(include_options_from_parent)
+            end
+
+            if serializer.respond_to?(:each)
+              parent[key] = to_many_relationships_for(serializer, opts)
+            else
+              parent[key] = to_one_relationship_for(serializer, opts)
+            end
+          end
+
+          parent
+        end
+
+        def relevant_included_associations_for(serializer, nested_associations = {})
+          if nested_associations.present?
+            serializer.associations.select do |association|
+              # nested_associations is a hash of:
+              #   key => nested association to include
+              nested_associations.key?(association.name)
+            end
+          else
+            serializer.associations
+          end
+        end
+
+        # add a singular relationship
+        # the options should always belong to the serializer
+        def to_one_relationship_for(serializer, options)
+          serialized_relationship = serialized_or_virtual_of(serializer, options)
+
+          nested_associations_to_include = options[:include]
+          if nested_associations_to_include.present?
+            serialized_relationship = add_resource_relationships(
+              serialized_relationship,
+              serializer,
+              nested_associations_to_include)
+          end
+
+          serialized_relationship
+        end
+
+        # add a many relationship
+        def to_many_relationships_for(serializer, options)
+          serialize_array(serializer) do |serialized_item, item_serializer|
+            nested_associations_to_include = options[:include]
+
+            if nested_associations_to_include.present?
+              serialized_item = add_resource_relationships(
+                serialized_item,
+                item_serializer,
+                nested_associations_to_include)
+            end
+
+            serialized_item
+          end
+        end
+
+        def serialize_array_without_root(serializer, options)
+          serializer.map { |s| FlattenJson.new(s).serializable_hash(options) }
+        end
+
+        # a virtual value is something that doesn't need a serializer,
+        # such as a ruby array, or any other raw value
+        def serialized_or_virtual_of(serializer, options)
+          if options[:virtual_value]
+            options[:virtual_value]
+          elsif serializer && serializer.object
+            resource_object_for(serializer)
+          end
+        end
+
+        def serialize_array(serializer)
+          serializer.map do |item|
+            serialized_item = resource_object_for(item)
+            serialized_item = yield(serialized_item, item) if block_given?
```

Comment:
This control flow is kinda obscure. Any way to rewrite it in a more easily understandable way, possibly avoiding `yield`?